skorch/probabilistic.py

```diff
@@ -0,0 +1,941 @@
+          e.g. 'cuda:0'.
+
+        Returns
+        -------
+        y_infer : list of gpytorch.distributions.Distribution
+          A list of distributions as generated by the module. The number of
+          elements in this list will depend on the sample size of X and the
+          batch size of the estimator.
+
+        """
+        y_infer = list(self.forward_iter(X, training=training, device=device))
+        return y_infer
+
+    def predict_proba(self, X):
+        raise AttributeError("'predict_proba' is not implemented for {}".format(
+            self.__class__.__name__
+        ))
+
+    def sample(self, X, n_samples, axis=-1):
+        """Return samples conditioned on input data.
+
+        The GP doesn't need to be fitted but it must be initialized.
+
+        X : input data
+          The samples where the GP is evaluated.
+
+        n_samples : int
+          The number of samples to return
+
+        axis : int (default=-1)
+          The concatenation axis of the samples. Since samples can come in
+          batches, they must be concatenated.
+
+        Returns
+        -------
+        samples : torch.Tensor
+          Samples from the posterior distribution.
+
+        """
+        self.check_is_fitted()
+        samples = [p.sample(torch.Size([n_samples])) for p in self.forward_iter(X)]
+        return torch.cat(samples, axis=axis)
+
+    def confidence_region(self, X, sigmas=2):
+        """Returns 2 standard deviations above and below the mean.
+
+        X : input data
+          The samples where the GP is evaluated.
+
+        sigmas : int (default=2)
+          The number of standard deviations of the region.
+
+        Returns
+        -------
+        lower : torch.Tensor
+          The lower end of the confidence region.
+
+        upper : torch.Tensor
+          The upper end of the confidence region.
+
+        """
+        nonlin = self._get_predict_nonlinearity()
+        lower, upper = [], []
+        for yi in self.forward_iter(X):
+            posterior = yi[0] if isinstance(yi, tuple) else yi
+            mean = posterior.mean
+            std = posterior.stddev
+            std = std.mul_(sigmas)
+            lower.append(nonlin(mean.sub(std)))
+            upper.append(nonlin(mean.add(std)))
+
+        lower = torch.cat(lower)
+        upper = torch.cat(upper)
+        return lower, upper
+
+    def __getstate__(self):
+        try:
+            return super().__getstate__()
+        except pickle.PicklingError as exc:
+            msg = ("This GPyTorch model cannot be pickled. The reason is probably this:"
+                   " https://github.com/pytorch/pytorch/issues/38137. "
+                   "Try using 'dill' instead of 'pickle'.")
+            raise pickle.PicklingError(msg) from exc
+
+
+class _GPRegressorPredictMixin:
+    """Mixin class that provides a predict method for GP regressors."""
+    def predict(self, X, return_std=False, return_cov=False):
```

Comment:
When `return_std is False`, consider turning the `gpytorch.settings.skip_posterior_variances` setting on -- this will avoid doing the math for the posterior variances altogether, which will save a great deal of compute.

Reply:
Good point, I will add this.